sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala

```diff
@@ -242,24 +272,48 @@ class UnivocityParser(
         new RuntimeException("Malformed CSV record"))
     } else {
       // When the length of the returned tokens is identical to the length of the parsed schema,
-      // we just need to convert the tokens that correspond to the required columns.
-      var badRecordException: Option[Throwable] = None
+      // we just need to:
+      //  1. Convert the tokens that correspond to the parsed schema.
+      //  2. Apply the pushdown filters to `parsedRow`.
+      //  3. Convert `parsedRow` to `requiredRow` by stripping non-required fields.
       var i = 0
+      val requiredSingleRow = requiredRow.head
       while (i < requiredSchema.length) {
+        requiredSingleRow.setNullAt(i)
+        i += 1
+      }
+
+      var skipValueConversion = false
+      var badRecordException: Option[Throwable] = None
+      i = 0
+      while (!skipValueConversion && i < parsedSchema.length) {
         try {
-          row(i) = valueConverters(i).apply(getToken(tokens, i))
+          val convertedValue = valueConverters(i).apply(getToken(tokens, i))
+          parsedRow(i) = convertedValue
+          if (csvFilters.skipRow(parsedRow, i)) {
+            skipValueConversion = true
+          } else {
+            val requiredIndex = parsedToRequiredIndex(i)
+            if (requiredIndex != -1) {
+              requiredSingleRow(requiredIndex) = convertedValue
+            }
+          }
         } catch {
           case NonFatal(e) =>
-            badRecordException = badRecordException.orElse(Some(e))
-            row.setNullAt(i)
+            badRecordException = Some(e)
+            skipValueConversion = true
         }
         i += 1
       }
-
-      if (badRecordException.isEmpty) {
-        row
+      if (skipValueConversion) {
+        if (badRecordException.isDefined) {
```

Comment:
how can we hit exception if `skipValueConversion` becomes true?

Reply:
Look at https://github.com/apache/spark/pull/26973/files#diff-c82e4b74d2a51fed29069745ce4f9e96R303-R304